README
NAME
    Error::Pure::Output::Tags::HTMLCustomPage - Error::Pure HTML output
    helper.

SYNOPSIS
     use Error::Pure::Output::Tags::HTMLCustomPage qw(err_pretty);

     err_pretty($tags_obj, $encoding, $content_type, $xml_version, $tags_structure_ar);

SUBROUTINES
    "err_pretty($tags_obj, $encoding, $content_type, $xml_version,
    $tags_structure_ar)"
             Helper routine use Tags $tags_obj object and print output to stdout.

EXAMPLE1
     use strict;
     use warnings;

     use Error::Pure::Output::Tags::HTMLCustomPage qw(err_pretty);
     use Tags::Output::Indent;

     # Tags object.
     my $tags = Tags::Output::Indent->new(
             'output_handler' => \*STDOUT,
             'auto_flush' => 1,
     );

     # Error.
     err_pretty($tags, 'utf-8', 'application/xhtml+xml', '1.0', [
             ['b', 'html'],
             ['b', 'head'],
             ['b', 'title'],
             ['d', 'Foo'],
             ['e', 'title'],
             ['e', 'head'],
             ['b', 'div'],
             ['d', 'Bar'],
             ['e', 'div'],
             ['e', 'html'],
     ]);

     # Output like:
     # Cache-Control: no-cache
     # Date: Wed, 03 Sep 2014 11:48:37 GMT
     # Content-Type: application/xhtml+xml
     #
     # <?xml version="1.0" encoding="utf-8" standalone="no"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html>
     #   <head>
     #     <title>
     #       Foo
     #     </title>
     #   </head>
     #   <div>
     #     Bar
     #   </div>
     # </html>

EXAMPLE2
     use strict;
     use warnings;

     use Error::Pure::Output::Tags::HTMLCustomPage qw(err_pretty);
     use Tags::Output::Raw;

     # Tags object.
     my $tags = Tags::Output::Raw->new(
             'output_handler' => \*STDOUT,
             'auto_flush' => 1,
     );

     # Error.
     err_pretty($tags, 'utf-8', 'application/xhtml+xml', '1.0', [
             ['b', 'html'],
             ['b', 'head'],
             ['b', 'title'],
             ['d', 'Foo'],
             ['e', 'title'],
             ['e', 'head'],
             ['b', 'div'],
             ['d', 'Bar'],
             ['e', 'div'],
             ['e', 'html'],
     ]);

     # Output like:
     # Cache-Control: no-cache
     # Date: Wed, 03 Sep 2014 11:54:37 GMT
     # Content-Type: application/xhtml+xml
     #
     # <?xml version="1.0" encoding="utf-8" standalone="no"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html><head><title>Foo</title></head><div>Bar</div></html>

DEPENDENCIES
    Exporter, HTTP::Headers::Fast, Readonly.

SEE ALSO
    Task::Error::Pure
        Install the Error::Pure modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/Error-Pure-Output-Tags-HTMLCusto
    mPage>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2014-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.05



Changes
0.05

0.04 2023-11-24T18:55:27+01:00
 - Fix dependencies in doc.
 - Fix use lib '.' in Makefile.PL.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Move bugtracker to github.
 - Rename example files to better names.
 - Update Module::Install to 1.21 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.03 2015-09-18T10:06:48+02:00
 - Add minimal version of Perl.
 - Add repository to doc.
 - Fix test dependencies.
 - Improve AUTHOR section in doc.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SEE ALSO section in doc.
 - Move author tests to xt/ directory.
 - Remove diacritics from author name.
 - Remove trailing whitespace.
 - Update Module::Install to 1.16 version.
 - Update copyright years.

0.02 2014-10-13T11:19:30+02:00
 - Add LICENSE file.

0.01 2014-09-03T13:56:45+02:00
 - First version.


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Error::Pure HTML output helper.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.03;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Error-Pure-Output-Tags-HTMLCustomPage';
readme_from 'HTMLCustomPage.pm';
recursive_author_tests('xt');
requires 'Exporter' => 0;
requires 'HTTP::Headers::Fast' => 0;
requires 'Readonly' => 0;
requires 'perl' => '5.8.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/'.
	'Error-Pure-Output-Tags-HTMLCustomPage/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/'.
	'Error-Pure-Output-Tags-HTMLCustomPage';
resources 'repository' => 'git://github.com/michal-josef-spacek/'.
	'Error-Pure-Output-Tags-HTMLCustomPage';
test_requires 'Test::NoWarnings' => 0;
test_requires 'Test::More' => 0;
tests_recursive;
version '0.05';

# Run.
WriteAll();


examples/ex1.pl
#!/usr/bin/env perl

# Pragmas.
use strict;
use warnings;

# Modules.
use Error::Pure::Output::Tags::HTMLCustomPage qw(err_pretty);
use Tags::Output::Indent;

# Tags object.
my $tags = Tags::Output::Indent->new(
        'output_handler' => \*STDOUT,
        'auto_flush' => 1,
);

# Error.
err_pretty($tags, 'utf-8', 'application/xhtml+xml', '1.0', [
        ['b', 'html'],
        ['b', 'head'],
        ['b', 'title'],
        ['d', 'Foo'],
        ['e', 'title'],
        ['e', 'head'],
        ['b', 'div'],
        ['d', 'Bar'],
        ['e', 'div'],
        ['e', 'html'],
]);

# Output:
# Cache-Control: no-cache
# Date: Wed, 03 Sep 2014 11:48:37 GMT
# Content-Type: application/xhtml+xml
# 
# <?xml version="1.0" encoding="utf-8" standalone="no"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html>
#   <head>
#     <title>
#       Foo
#     </title>
#   </head>
#   <div>
#     Bar
#   </div>
# </html>

examples/ex2.pl
#!/usr/bin/env perl

# Pragmas.
use strict;
use warnings;

# Modules.
use Error::Pure::Output::Tags::HTMLCustomPage qw(err_pretty);
use Tags::Output::Raw;

# Tags object.
my $tags = Tags::Output::Raw->new(
        'output_handler' => \*STDOUT,
        'auto_flush' => 1,
);

# Error.
err_pretty($tags, 'utf-8', 'application/xhtml+xml', '1.0', [
        ['b', 'html'],
        ['b', 'head'],
        ['b', 'title'],
        ['d', 'Foo'],
        ['e', 'title'],
        ['e', 'head'],
        ['b', 'div'],
        ['d', 'Bar'],
        ['e', 'div'],
        ['e', 'html'],
]);

# Output like:
# Cache-Control: no-cache
# Date: Wed, 03 Sep 2014 11:54:37 GMT
# Content-Type: application/xhtml+xml
# 
# <?xml version="1.0" encoding="utf-8" standalone="no"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html><head><title>Foo</title></head><div>Bar</div></html>

HTMLCustomPage.pm
package Error::Pure::Output::Tags::HTMLCustomPage;

use base qw(Exporter);
use strict;
use warnings;

use HTTP::Headers::Fast;
use Readonly;

# Constants.
Readonly::Array our @EXPORT_OK => qw(err_pretty);
Readonly::Scalar my $EMPTY_STR => q{};
Readonly::Scalar my $SPACE => q{ };

our $VERSION = 0.05;

# Pretty print.
sub err_pretty {
	my ($tags_obj, $encoding, $content_type, $xml_version,
		$tags_structure_ar) = @_;

	# Reset.
	$tags_obj->reset;

	my $header = HTTP::Headers::Fast->new;
	$header->header(
		'Cache-Control' => 'no-cache',
		'Content-Type' => $content_type,
	);
	$header->date(time);
	$tags_obj->put(['r', $header->as_string."\n"]);

	# Debug from Tags object.
	my $debug = $tags_obj->{'set_indent'};

	# XML tag.
	$tags_obj->put(
		['i', 'xml', 'version="'.$xml_version.'" encoding="'.$encoding.
			'" standalone="no"'],
	);
	if ($debug) {
		$tags_obj->put(['r', "\n"]);
	}

	# DTD.
	my @dtd = (
		'<!DOCTYPE html',
		'PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"',
		'"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">',
	);
	my ($separator, $linebreak) = ($SPACE, $EMPTY_STR);
	if ($debug) {
		$separator = "\n".$tags_obj->{'next_indent'};
		$linebreak = "\n";
	}
	$tags_obj->put(['r', (join $separator, @dtd).$linebreak]);

	# Main page.
	my @tmp = @{$tags_structure_ar};
	$tags_obj->put(@tmp);

	# Ok.
	return 1;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Error::Pure::Output::Tags::HTMLCustomPage - Error::Pure HTML output helper.

=head1 SYNOPSIS

 use Error::Pure::Output::Tags::HTMLCustomPage qw(err_pretty);

 err_pretty($tags_obj, $encoding, $content_type, $xml_version, $tags_structure_ar);

=head1 SUBROUTINES

=over 8

=item C<err_pretty($tags_obj, $encoding, $content_type, $xml_version, $tags_structure_ar)>

 Helper routine use Tags $tags_obj object and print output to stdout.

=back

=head1 EXAMPLE1

=for comment filename=error_as_html_indent.pl

 use strict;
 use warnings;

 use Error::Pure::Output::Tags::HTMLCustomPage qw(err_pretty);
 use Tags::Output::Indent;

 # Tags object.
 my $tags = Tags::Output::Indent->new(
         'output_handler' => \*STDOUT,
         'auto_flush' => 1,
 );

 # Error.
 err_pretty($tags, 'utf-8', 'application/xhtml+xml', '1.0', [
         ['b', 'html'],
         ['b', 'head'],
         ['b', 'title'],
         ['d', 'Foo'],
         ['e', 'title'],
         ['e', 'head'],
         ['b', 'div'],
         ['d', 'Bar'],
         ['e', 'div'],
         ['e', 'html'],
 ]);

 # Output like:
 # Cache-Control: no-cache
 # Date: Wed, 03 Sep 2014 11:48:37 GMT
 # Content-Type: application/xhtml+xml
 #
 # <?xml version="1.0" encoding="utf-8" standalone="no"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html>
 #   <head>
 #     <title>
 #       Foo
 #     </title>
 #   </head>
 #   <div>
 #     Bar
 #   </div>
 # </html>

=head1 EXAMPLE2

=for comment filename=error_as_html_raw.pl

 use strict;
 use warnings;

 use Error::Pure::Output::Tags::HTMLCustomPage qw(err_pretty);
 use Tags::Output::Raw;

 # Tags object.
 my $tags = Tags::Output::Raw->new(
         'output_handler' => \*STDOUT,
         'auto_flush' => 1,
 );

 # Error.
 err_pretty($tags, 'utf-8', 'application/xhtml+xml', '1.0', [
         ['b', 'html'],
         ['b', 'head'],
         ['b', 'title'],
         ['d', 'Foo'],
         ['e', 'title'],
         ['e', 'head'],
         ['b', 'div'],
         ['d', 'Bar'],
         ['e', 'div'],
         ['e', 'html'],
 ]);

 # Output like:
 # Cache-Control: no-cache
 # Date: Wed, 03 Sep 2014 11:54:37 GMT
 # Content-Type: application/xhtml+xml
 #
 # <?xml version="1.0" encoding="utf-8" standalone="no"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html><head><title>Foo</title></head><div>Bar</div></html>

=head1 DEPENDENCIES

L<Exporter>,
L<HTTP::Headers::Fast>,
L<Readonly>.

=head1 SEE ALSO

=over

=item L<Task::Error::Pure>

Install the Error::Pure modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Error-Pure-Output-Tags-HTMLCustomPage>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2014-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.05

=cut
